packages/@lwc/template-compiler/src/codegen/expression.ts

```diff
@@ -0,0 +1,169 @@
+ *    {(foo) => foo && bar} -> {($cmp.foo) => foo && $cmp.bar}
+ *    {(foo) => foo && bar} -> {($cmp.foo) => $cmp.foo && $cmp.bar}
+ *
+ * Instead, the scopes are respected:
+ *    {(foo) => foo && $cmp.bar}
+ *
+ * Similar checks occur for local identifiers introduced via for:each or similar.
+ */
+export function bindComplexExpression(
+    expression: ComplexExpression,
+    codeGen: CodeGen
+): t.Expression {
+    const expressionScopes = new ExpressionScopes();
+    walk(expression, {
+        enter(node, _parent) {
+            if (t.isArrowFunctionExpression(node)) {
+                expressionScopes.enterScope(node);
+            }
+        },
+
+        leave(node, parent) {
+            if (t.isArrowFunctionExpression(node)) {
+                expressionScopes.exitScope(node);
+            } else if (
+                parent !== null &&
+                t.isIdentifier(node) &&
+                // Acorn parses `undefined` as an Identifier.
+                node.name !== 'undefined' &&
+                !(t.isMemberExpression(parent) && parent.property === node) &&
+                !(t.isProperty(parent) && parent.key === node) &&
+                !codeGen.isLocalIdentifier(node) &&
+                !expressionScopes.isScopedToExpression(node)
+            ) {
+                this.replace(t.memberExpression(t.identifier(TEMPLATE_PARAMS.INSTANCE), node));
+            }
+        },
+    });
+
+    return expression as t.Expression;
+}
+
+/**
+ * Track the variables that come in and out of scope in various parts of a
+ * template expression. Arrow functions can return arrow functions, which can lead to
+ * variable shadowing, which needs to be handled correctly.
+ */
+class ExpressionScopes {
+    variableShadowingCount = new Map<VariableName, VariableShadowingMultiplicity>();
+    arrowFnVariables = new Map<t.ArrowFunctionExpression, Set<VariableName>>();
+
+    enterScope(node: t.ArrowFunctionExpression) {
+        const variableNamesIntroduced: VariableNames = new Set();
+        for (const param of node.params) {
+            collectParams(param, variableNamesIntroduced);
+        }
+        for (const varName of variableNamesIntroduced) {
+            this.variableShadowingCount.set(
+                varName,
+                (this.variableShadowingCount.get(varName) ?? 0) + 1
+            );
+        }
+        this.arrowFnVariables.set(node, variableNamesIntroduced);
+    }
+
+    exitScope(node: t.ArrowFunctionExpression) {
+        const varNames = this.arrowFnVariables.get(node);
+        if (varNames) {
+            for (const varName of varNames) {
+                this.variableShadowingCount.set(
+                    varName,
+                    this.variableShadowingCount.get(varName)! - 1
+                );
+            }
+        }
+    }
+
+    // If a variable was introduced in an arrow function and is still
+    // in scope, return true. Otherwise, return false.
```

Comment:
I take it we only allow `const`/`let`, not `var`? (Or heaven forbid, global variable assignment? 🙂)

Reply:
Statements are disallowed, including variable declarations. This only applies to variables introduces as arrow fn params.

Reply:
Clarified this in the comment.